ACS-3183 POST rules, GET, PUT and DELETE rule API definition.

tpage-alfresco opened the pull request:
Initial spec for Rules CRUD. More PRs to follow, eg. to add link/unlink rulesets etc.

Also some small changes to the GET rules API, the largest being the introduction
of a new type for action definitions (one that doesn't include a target node id).

tpage-alfresco commented on src/main/webapp/definitions/alfresco-core.yaml:
```diff
@@ -8472,7 +8472,48 @@ paths:
+        **Note:** this endpoint is available in Alfresco 7.3 and newer versions.
 
+        Creates a rule on folder **nodeId**.
+      operationId: createComment
```
This still mentions comments

krdabrowski commented on src/main/webapp/definitions/alfresco-core.yaml:
```diff
@@ -10758,12 +10809,68 @@ definitions:
         items:
-          $ref: '#/definitions/ActionDefinition'
+          $ref: '#/definitions/ActionBodyExecTemplate'
+  RuleBody:
```
The `RuleBody` is almost the same as `Rule`, isn't it possible to use one object in both cases? Or introduce inheritance?

tpage-alfresco replied:
This wasn't possible last time I checked.  There's syntax for `allOf`, `anyOf` and `oneOf` which allow this sort of reuse, but IIRC there was some reason we couldn't use it.  @montgolfiere Do you remember any more about this issue?

tpage-alfresco commented:
Added API definitions for GET/PUT/DELETE a single rule and updated everything to include the concept of a ruleset (so now we have the three level path `/nodes/{nodeId}/rulesets/{rulesetId}/rules/{ruleId}`).

montgolfiere commented on src/main/webapp/definitions/alfresco-core.yaml:
```diff
@@ -8435,7 +8447,7 @@ paths:
           description: |
             **actionDefinitionId** or **targetId** does not exist
-  '/nodes/{nodeId}/rules':
+  '/nodes/{nodeId}/rulesets/{rulesetId}/rules':
```
If this is -default- (or similar) then should be stated in each method description (alongside folder "nodeId").

montgolfiere reviewed (commented):
Thanks @tpage-alfresco. It looks like this coming together nicely. I've a few minor initial comments

We also will need to add support for:

- re-order rules
- linked rulesets (link , unlink, view - who I link to, who links to me)
- disable / enable rule inheritance

I realise we may decide to add incrementally via separate PRs (eg. see https://github.com/Alfresco/rest-api-explorer/pull/159 ).

Also consider new features / stretch goals such as:

- OR conditions (as per composite nested grouping)
- is null / is not null
- copy rule (presumably to another folder)

montgolfiere reviewed (approved):
As discussed, I think we can merge this initial Rules CRUD spec proposal on the basis that:

- this is still experimental ( & hence subject to change ) targeting currently planned ACS Content Repo 7.3.0
- we will have follow-on PRs for other required APIs (eg. link / unlink rulesets as per Draft https://github.com/Alfresco/rest-api-explorer/pull/159) and any other identified gaps &/or new features (see also https://github.com/Alfresco/rest-api-explorer/pull/161#pullrequestreview-1027420129 )
- we will keep up-to-date as we build out the API impl (eg. [first step](https://github.com/Alfresco/alfresco-community-repo/pull/1193)) as well as end-to-end API test scenarios and client usage (eg. ACA/ADW UIs)
- we will review with ACS architects (once we have the ability to demonstrate most/all end-to-end scenarios)
